### PR TITLE
Fix and add a test for failure store with Incremental bulk

### PR DESCRIPTION
**Pull request by ankikuma**
When a document is rejected because of indexing pressure, it should not be redirected to the failure store.

The failure store is not meant to be a dead letter queue - it’s a best effort storage location for documents that cannot be ingested because there is some kind of fault in their shape or content, this way a user can fix them. 

In the case of indexing pressure there is nothing wrong with the document itself. In this PR we fix the redirection to the failure store and we add an integration test to test the interaction of the failure store and incremental bulk's short circuit failure feature.

Closes ES-9577.

### CODE REVIEW

File: modules/data-streams/src/internalClusterTest/java/org/elasticsearch/datastreams/FailureStoreMetricsWithIncrementalBulkIT.java
Comment (Tim-Brooks) [inline modules/data-streams/src/internalClusterTest/java/org/elasticsearch/datastreams/FailureStoreMetricsWithIncrementalBulkIT.java:170]: Just so I understand this correctly:

If an indexing operation fails and there is a failure store configured will then try to store that failure in the failure store. If that operation succeeds, then the indexing operations is indicated to the user as "successful"?

As long as that is the expected behavior this mitigation looks good to me.

File: modules/data-streams/src/internalClusterTest/java/org/elasticsearch/datastreams/FailureStoreMetricsWithIncrementalBulkIT.java
Comment (Tim-Brooks) [inline modules/data-streams/src/internalClusterTest/java/org/elasticsearch/datastreams/FailureStoreMetricsWithIncrementalBulkIT.java:167]: Java variables should be camel case. `docsInFS`

File: modules/data-streams/src/internalClusterTest/java/org/elasticsearch/datastreams/FailureStoreMetricsWithIncrementalBulkIT.java
Comment (Tim-Brooks) [inline modules/data-streams/src/internalClusterTest/java/org/elasticsearch/datastreams/FailureStoreMetricsWithIncrementalBulkIT.java:166]: Java variables should be camel case. `docsRedirectedToFs`

**Review by Tim-Brooks — COMMENTED** [review]
LGTM if someone from the data management team confirms this is the expected behavior and what they were looking to test.

**Comment by ankikuma** [comment]
@gmarouli @nielsbauman Could you please take a look at this test. This test is added as a follow up to the discussion [here](https://elastic.slack.com/archives/CE8P96STG/p1726598358836249). The test behavior is as follows:

1. We first try to index the document into the datastream.
2. When the document is rejected (due to indexing pressure in our testcase), the document is redirected to the failure store.
3. If the failure store's backing index is on the same node as the datastream's backing index, the same indexing pressure will apply to the failure store and the request will in fact fail because we are unable to index the doc into the failure store. In this case the getFailureStoreStatus() for this request will be FAILED.
4. But if the failure store's backing index is on a different node, the request will succeed because the doc is successfully stored in the failure store. In this case the getFailureStoreStatus() for this request will be USED.

Based on the slack thread linked above, we don't want to use the failure store to store docs that were rejected due to a resource level failure. But my testcase shows that we do. So perhaps we need a follow up change to the failure store code.

**Comment by nielsbauman** [comment]
Hi @ankikuma, I'm currently not on the failure store project anymore so I unfortunately won't be able to have a look at this. I'll leave you in the good hands of @gmarouli :)

File: modules/data-streams/src/internalClusterTest/java/org/elasticsearch/datastreams/FailureStoreMetricsWithIncrementalBulkIT.java
Comment (ankikuma) [reply modules/data-streams/src/internalClusterTest/java/org/elasticsearch/datastreams/FailureStoreMetricsWithIncrementalBulkIT.java:170]: Yes your understanding is correct @Tim-Brooks. If we successfully index into the failure store, the indexing operation is considered successful, even though the operation failed to index into the original index.

File: modules/data-streams/src/internalClusterTest/java/org/elasticsearch/datastreams/FailureStoreMetricsWithIncrementalBulkIT.java
Comment (ankikuma) [reply modules/data-streams/src/internalClusterTest/java/org/elasticsearch/datastreams/FailureStoreMetricsWithIncrementalBulkIT.java]: Done.

File: modules/data-streams/src/internalClusterTest/java/org/elasticsearch/datastreams/FailureStoreMetricsWithIncrementalBulkIT.java
Comment (gmarouli) [reply modules/data-streams/src/internalClusterTest/java/org/elasticsearch/datastreams/FailureStoreMetricsWithIncrementalBulkIT.java:170]: Hi @ankikuma & @Tim-Brooks ,

> As long as that is the expected behavior this mitigation looks good to

I do not think this is expected behaviour. 

I am afraid I mislead @ankikuma during our last chat.

> If we successfully index into the failure store, the indexing operation is considered successful, even though the operation failed to index into the original index.

This is correct, and it explains the behaviour that the code exhibits. But if I remember correctly the conversations we looked into, the purpose of the failure store is to store failures as a result from a user misconfiguration and not technical limitations/failures. 

Considering this, I would expect that this type of failure should not be redirected to the failure store and it should result in a failed response.

As a way forward, I see two options depending on the scope of this work:
- If the purpose of this test is to check that incremental indexing and failure store work as expected, I would say that we need to fix the bug that this test has unearthed.
- If the purpose is to only add a test to cover the failure store and incremental bulk indexing working together, we should write the test to work as expected and then open a bug and mute this test.

Preferably, I would prefer the first but I do not fully know the scope of this work, so I would like to offer an alternative as well.

**Review by gmarouli — COMMENTED** [review]
Hi @ankikuma I hold off the review until I understand better the scope of this PR. But it looks good in general :). 

Thank you for including the failure store in your test suite!

File: modules/data-streams/src/internalClusterTest/java/org/elasticsearch/datastreams/FailureStoreMetricsWithIncrementalBulkIT.java
Comment (ankikuma) [reply modules/data-streams/src/internalClusterTest/java/org/elasticsearch/datastreams/FailureStoreMetricsWithIncrementalBulkIT.java:170]: @gmarouli The scope of this test was to just test the interaction of incremental bulk failures caused by the short circuit feature with failure store, based on the comment from @nielsbauman [here](https://github.com/elastic/elasticsearch/pull/113044#pullrequestreview-2312162329). And in this original comment , it looks like we expected the short circuited requests to got to the failure store.

However, as this [discussion](https://elastic.slack.com/archives/CE8P96STG/p1726684093864779?thread_ts=1726598358.836249&cid=CE8P96STG) suggests, James Baiera indicated that if failures occurred due to resource constraints, they should not go to the failure store.

Now a short circuit failure is triggered due to a previous failure. It just so happens that in this test we simulate that failure based on indexing pressure. I am not sure how one would distinguish between failures caused by resource constraints vs. other types of failures (for the benefit of the failure store).

File: modules/data-streams/src/internalClusterTest/java/org/elasticsearch/datastreams/FailureStoreMetricsWithIncrementalBulkIT.java
Comment (gmarouli) [inline modules/data-streams/src/internalClusterTest/java/org/elasticsearch/datastreams/FailureStoreMetricsWithIncrementalBulkIT.java:88]: @ankikuma , I thought this test is pretty targeted to failure store so we could merge the template and data stream creation in one method.

File: modules/data-streams/src/internalClusterTest/java/org/elasticsearch/datastreams/FailureStoreMetricsWithIncrementalBulkIT.java
Comment (gmarouli) [inline modules/data-streams/src/internalClusterTest/java/org/elasticsearch/datastreams/FailureStoreMetricsWithIncrementalBulkIT.java:97]: @ankikuma I renamed this to `successfullyStored` because `hits` confused me since I have associated it with search results. I also did not see the need to use an atomic.

File: modules/data-streams/src/internalClusterTest/java/org/elasticsearch/datastreams/FailureStoreMetricsWithIncrementalBulkIT.java
Comment (gmarouli) [inline modules/data-streams/src/internalClusterTest/java/org/elasticsearch/datastreams/FailureStoreMetricsWithIncrementalBulkIT.java:111]: @ankikuma I abstracted finding the node of the primary shard in one method because I thought the retrieval did not add much to the flow of the test.

File: modules/data-streams/src/internalClusterTest/java/org/elasticsearch/datastreams/FailureStoreMetricsWithIncrementalBulkIT.java
Comment (gmarouli) [inline modules/data-streams/src/internalClusterTest/java/org/elasticsearch/datastreams/FailureStoreMetricsWithIncrementalBulkIT.java:216]: @ankikuma I cleaned up the measurements assertion here a bit, not all of them were used and I found this structure a bit easier to follow. 

The way I understand it, we provide the observed telemetry, the metric and the data streams we are interested in and the expected value, then it will sum the values up and verify.

The previous one was also correct, but I find this a bit more intuitive while before I got a bit lost why do we need the size of the list ;).

I am also using at every point when we check the measurements, even when the measurements would be empty because I also find the symmetry easier to read.

**Review by gmarouli — COMMENTED** [review]
Hi @ankikuma , as I started working on changing the test to match the fix, I noticed a few things that were confusing me a bit. I added an explanation for every change I performed that was not related with the actual fix. 

This is still your PR so if you find that something does not suit you feel free to revert or improve.

**Comment by gmarouli** [comment]
@elasticmachine update branch

**Comment by ankikuma** [comment]
@jbaiera could you please review this test to see if it is aligned with the expected behavior of the failure store. Thank you Mary for making the changes!

**Review by jbaiera — APPROVED** [review]
The code changes and line of thinking LGTM.

✅ assuming green CI

**Pull request merged**
